### ISSUE TITLE
hipfire-quantize --kmap-dense --kmap-mode 2 (mixed MQ4/MQ6 dense): silent NaN logits at runtime

### ISSUE THREAD

**Issue by unverbraucht** [Kaden-Schutt/hipfire#249]
## Summary

`hipfire-quantize --format mq4 --kmap-dense --kmap-mode 2` produces a structurally well-formed model file (kmap mode 2 = "typed": `mlp.down_proj` + `attn_v` promoted to MQ6G256, rest stays MQ4G256), but the dense forward pass produces all-NaN logits at runtime. The eval runs to completion with no crash — only `slice-mean KLD = 0.000000  mean NLL = NaN  PPL = NaN` reveals the failure.

Same family as #179 and #209: runtime dispatch matchers that don't enumerate all valid tensor-format combinations. The previous two issues were MQ3-in-MoE matcher gaps; this one is **mixed MQ4/MQ6 in a single dense layer** (`w_gate=MQ4G256, w_down=MQ6G256`).

## Reproducer

```bash
# Quantize 9B with kmap-dense + typed mode 2
hipfire-quantize \
  --input /path/to/Qwen3.5-9B \
  --output ~/.hipfire/models/qwen3.5-9b.mq4-kmd2 \
  --format mq4 --kmap-dense --kmap-mode 2
```

The quantize log shows the intended per-layer mixture (verified on commit `1b90a663`):

```
MQ4G256: model.language_model.layers.8.linear_attn.in_proj_qkv.weight
MQ4G256: model.language_model.layers.8.linear_attn.out_proj.weight
MQ6G256: model.language_model.layers.8.mlp.down_proj.weight
MQ4G256: model.language_model.layers.8.mlp.gate_proj.weight
MQ4G256: model.language_model.layers.8.mlp.up_proj.weight
```

(In Qwen 3.5 dense, `attn_v` is fused into `in_proj_qkv` so the substring match in `kmap_resolve_mode` at `crates/hipfire-quantize/src/main.rs:2191` only catches `down_proj` — `attn_v` promotion is effectively a no-op for this arch. Independent docs question; not part of this bug.)

Eval (any prompt, any KV mode):

```
eval_hipfire: slice-mean KLD = 0.000000  mean NLL = NaN  PPL = NaN
```

The KLD == 0 is misleading — all logits are NaN, so every reference bin is "equally wrong" and the per-token KLD reduces to 0. Without the NLL/PPL check this would look like a perfect-fidelity result.

Same recipe with **uniform** format works:
- `--format mq4` (no `--kmap-dense`): correct (Mean KLD ≈ 0.25 on 9B at n=512)
- `--format mq6` (no `--kmap-dense`): correct (Mean KLD ≈ 0.057 on 9B at n=20)

Only the **mixed** `(MQ4 gate/up, MQ6 down)` configuration NaNs.

## Suspected cause (needs runtime audit)

Dense FFN body matchers at `crates/hipfire-arch-qwen35/src/qwen35.rs:4628-4715` correctly split gate/up dispatch (keyed on `w_gate.gpu_dtype`) from down dispatch (keyed on `w_down.gpu_dtype`). When the two formats disagree, each side picks its own kernel — but the activation handoff between them is not necessarily compatible. Candidate root causes (in order of likelihood):

1. **AWQ-scale mismatch across formats.** `fused_rmsnorm_rotate_mq_batched_for(..., &layer.w_gate, ...)` at line 4639 reads the AWQ scale from `w_gate`. The activations entering `gemm_gate_up_*` are pre-scaled by `1/awq_scale(w_gate)`. The post-SwiGLU activations entering `gemm_*_residual(w_down, …)` are expected to be pre-scaled by `1/awq_scale(w_down)`. When the two weights were quantized **independently with different format-specific AWQ scale derivations**, the cancellation no longer holds; the silu_mul output enters w_down with the wrong scale and overflows somewhere in the f16 path.

2. **FWHT direction/normalization difference between MQ4 and MQ6 kernels.** `fused_silu_mul_rotate_mq_batched` at line 4701 is called whenever `w_down_is_mq` is true, but it doesn't take the actual `gpu_dtype` — it just rotates. If the MQ4 vs MQ6 kernel families disagree on FWHT normalization convention or rotation direction, the rotated activation hits `gemm_hfq6g256_residual` in the wrong basis.

3. **gate_up output-range mismatch.** `gemm_gate_up_hfq4g256` (entered via the fall-through `else` for MQ4 at line 4677) may produce intermediates in a different numeric range than what `gemm_gate_up_hfq6g256` would, in a way that's safe when feeding the matching `w_down` kernel but explodes through silu_mul before the MQ6 down.

I haven't bisected which is the actual culprit — the audit pattern from #179 / #209 (read each dispatch matcher, trace activation ranges through the mixed case, compare to the all-MQ4 and all-MQ6 cases at the same prompt) should find it quickly.

## Why this isn't covered by existing matchers

The dense-FFN matchers in `qwen35.rs:4628-4715` look correct at first glance — they DO key `w_down_*` predicates off `layer.w_down.gpu_dtype` independently from `w_gate.gpu_dtype`. So this isn't a "missing dtype in matches!" gap like #179. The bug is one level deeper: the **kernels themselves aren't designed to interoperate across format families**, because no shipping model exercised the mixed configuration until `--kmap-dense --kmap-mode 2` produced one.

The dense LA/FA body wo dispatch (lines 4197-4382) probably has the same latent issue if you mix `wqkv` and `wo` formats — the kmd2 mode doesn't promote those (only `down_proj` got promoted in practice for Qwen 3.5 dense), so it's not exercised here, but a future kmap mode that promotes attention weights would hit it.

## Failure mode

Identical class to the MQ8 silent-corruption documented in `docs/perf-checkpoints/2026-05-06-mq8-runtime-dispatch-audit.md`: the model loads, prefill runs at normal speed (215 tok/s on gfx1100), no panic, no `nan_check` trip — only the downstream KLD/PPL reveals NaN logits. A future `--kmap-dense --kmap-mode 2` user without a KLD pipeline would get garbage tokens at the bench-measured speed.

## Workaround / current state

For the current empirical work, **uniform-format quantization is competitive**:

- 9B uniform `--format mq6` + `HIPFIRE_QUANTIZE_CONV_Q8=1` → Mean KLD 0.0568 at n=20, q8-KV (matches llama.cpp UD-Q4_K_XL at +1.2 bpw)
- 9B uniform `--format mq4` + `HIPFIRE_QUANTIZE_CONV_Q8=1` → Mean KLD 0.2501 at n=512, q8-KV

So `--kmap-dense --kmap-mode 2` is not blocking deliverables — but the "typed promotion" idea is real (down_proj is the highest-error linear in every per-layer KLD decomposition we've measured), and getting it working would be a free quality win at +0.5 bpw.

## Suggested fix

Three options, in increasing scope:

1. **Refuse the mixed config at load time.** Add a uniform-format check to the dense layer loader (mirror `moe_ffn_all_mq4` for the dense case) that errors with a clear message: "kmap-dense mode 2 produces mixed gate/down formats, currently unsupported — use `--format mq6` (uniform) instead." Low effort, prevents silent corruption.

2. **Audit and fix the mixed-format activation handoff.** Trace the AWQ-scale, FWHT, and output-range invariants between `gemm_gate_up_{hfq4g256,hfq6g256}` and `gemm_{hfq4g256,hfq6g256}_residual`. Patch the kernels or insert a scale-bridging step in the silu_mul_rotate path. Higher effort but unlocks per-tensor format tiering generally.

3. **Document `--kmap-dense` as uniform-format only.** Cheapest — but the typed-promotion ideal is good enough that a future contributor will rediscover this bug.

I'd land (1) as a guard now and file (2) as a separate "make per-tensor format tiering work" tracking issue if anyone wants to pick it up.

## Coverage gate

`scripts/coherence-gate.sh` doesn't currently exercise any `--kmap-dense` model, so this regression is invisible to the gate. Once the load-time guard or the kernel fix lands, adding a `qwen3.5-9b.mq4-kmd2` row would prevent future regressions.

## Related

- #179 — MQ3 missing from MoE-batched matchers (same dispatch-audit class)
- #209 — MQ3 + MoE batched prefill (followup tracking issue, same class)
- `docs/perf-checkpoints/2026-05-06-mq8-runtime-dispatch-audit.md` — MQ8 silent-corruption audit (same failure pattern: load OK, run OK, NaN logits)
- `docs/plans/kld-measurements-master.md` §5 — full empirical context for the kmd2 NaN finding alongside MQ6 / MQ4-Lloyd / HF4 results

(Filed via Claude Code agent on Kevin Read's behalf — issue text reviewed before submit. Bug observed on commit `1b90a663` on branch `feat/mq-v2-quant-format`, gfx1100.)

**Comment by unverbraucht** [comment]
Fixed.
